packages/shared/src/components/tabs/TabContainer.tsx

```diff
@@ -84,22 +87,25 @@ export function TabContainer<T extends string = string>({
   });
 
   const currentActive = controlledActive ?? active;
-  const onClick = (label: T) => {
+  const onClick: TabListProps['onClick'] = (label: T, event) => {
     const child = children.find((c) => c.props.label === label);
     setActive(label);
-    onActiveChange?.(label);
-
-    setTimeout(() => {
-      if (shouldFocusTabOnChange && containerRef?.current) {
-        const [firstChild] = containerRef.current.children;
-        if (firstChild instanceof HTMLElement) {
-          firstChild.focus();
+    const shouldChange = onActiveChange?.(label, event);
+
+    // evaluate !== false due to backwards compatibility with implementations that return undefined
+    if (shouldChange !== false) {
```

Comment:
Added support where we can return `false` from `onActiveChange` which will cancel any tab navigation (like `event.preventDefault` for native clicks.